crates/rust-analyzer/src/main_loop/handlers.rs

```diff
@@ -831,13 +843,23 @@ pub fn handle_code_lens(
 
     let mut lenses: Vec<CodeLens> = Default::default();
 
+    let cargo_spec = CargoTargetSpec::for_file(&world, file_id)?;
     // Gather runnables
     for runnable in world.analysis().runnables(file_id)? {
         let title = match &runnable.kind {
             RunnableKind::Test { .. } | RunnableKind::TestMod { .. } => "▶️\u{fe0e}Run Test",
             RunnableKind::DocTest { .. } => "▶️\u{fe0e}Run Doctest",
             RunnableKind::Bench { .. } => "Run Bench",
-            RunnableKind::Bin => "Run",
+            RunnableKind::Bin => {
+                // Do not suggest binary run on other target than binary
+                match &cargo_spec {
+                    Some(spec) => match spec.target_kind {
+                        TargetKind::Bin => "Run",
+                        _ => continue,
+                    },
+                    None => continue,
```

Comment:
`cargo_spec` is None, when using `rust-analyzer.json`  instead of `cargo metadata`.

Reply:
Ok so I should also return "Run" in that case ?

Reply:
Yes

Reply:
tbh, for `rust-projec.json` I am not sure if runnables are at all useful. SO, for me the behavior there doesn't really matter